internal/controllers/machineset/machineset_controller_status.go

```diff
@@ -0,0 +1,278 @@
+		var message string
+		if referencesMessage != "" {
+			message = fmt.Sprintf("ScalingUp would fail because %s", referencesMessage)
+		}
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:    clusterv1.MachineSetScalingUpV1Beta2Condition,
+			Status:  metav1.ConditionFalse,
+			Reason:  clusterv1.MachineSetNotScalingUpV1Beta2Reason,
+			Message: message,
+		})
+		return
+	}
+
+	message := fmt.Sprintf("ScalingUp from %d to %d replicas", gotReplicas, desiredReplicas)
+	if referencesMessage != "" {
+		message = fmt.Sprintf("%s is blocked because %s", message, referencesMessage)
+	}
+	// Scaling up.
+	v1beta2conditions.Set(ms, metav1.Condition{
+		Type:    clusterv1.MachineSetScalingUpV1Beta2Condition,
+		Status:  metav1.ConditionTrue,
+		Reason:  clusterv1.MachineSetScalingUpV1Beta2Reason,
+		Message: message,
+	})
+}
+
+func setScalingDownCondition(ms *clusterv1.MachineSet, machines []*clusterv1.Machine, bootstrapObjectNotFound, infrastructureObjectNotFound bool, owner *clusterv1.MachineDeployment, machinesStaleDeleting []string) {
+	// If we got unexpected errors in listing the machines (this should happen rarely), surface them
+	if machines == nil {
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:    clusterv1.MachineSetScalingDownV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  clusterv1.MachineSetScalingDownErrorV1Beta2Reason,
+			Message: "Please check controller logs for errors",
+		})
+		return
+	}
+
+	// Surface if .spec.replicas is not yet set (this should happen rarely).
+	// This could e.g. be the case when using autoscaling with MachineSets.
+	if ms.Spec.Replicas == nil {
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:   clusterv1.MachineSetScalingDownV1Beta2Condition,
+			Status: metav1.ConditionUnknown,
+			Reason: clusterv1.MachineSetWaitingForReplicasToBeSetV1Beta2Reason,
+		})
+		return
+	}
+
+	desiredReplicas := *ms.Spec.Replicas
+
+	// Deletion is equal to 0 desired replicas.
+	if !ms.DeletionTimestamp.IsZero() {
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:    clusterv1.MachineSetScalingDownV1Beta2Condition,
+			Status:  metav1.ConditionTrue,
+			Reason:  clusterv1.MachineSetScalingDownV1Beta2Reason,
+			Message: "MachineSet is getting deleted",
+		})
+		return
+	}
+
+	// Scaling down.
+	if int32(len(machines)) > (desiredReplicas) {
+		messages := []string{fmt.Sprintf("ScalingDown from %d to %d replicas", len(machines), desiredReplicas)}
+
+		if len(machinesStaleDeleting) > 0 {
+			messages = append(messages, fmt.Sprintf("Machines stuck in deletion for more than 30 minutes: %s", clog.StringListToString(machinesStaleDeleting)))
+		}
+
+		// Scaling is blocked when the bootstrap or infrastructure objects do not exist.
+		if bootstrapObjectNotFound || infrastructureObjectNotFound {
+			messages[0] = fmt.Sprintf("%s is blocked because %s", messages[0], calculateMissingReferencesMessage(bootstrapObjectNotFound, infrastructureObjectNotFound))
+		}
```

Comment:
might be move this above, right after we compute messages[0] (and possibly use a var instread of message [0]
```suggestion
		message := "..."
		if .. { message += "..."}
		messages = []string{message}
```